ext/stackprof/stackprof.c

```diff
@@ -370,6 +454,53 @@ stackprof_results(int argc, VALUE *argv, VALUE self)
     st_free_table(_stackprof.frames);
     _stackprof.frames = NULL;
 
+    if (_stackprof.tag_strings_len > 0) {
+        tag_strings = rb_ary_new_capa(_stackprof.tag_strings_len);
+        for (size_t n = 0; n < _stackprof.tag_strings_len; n++) {
+            rb_ary_push(tag_strings, rb_str_new_cstr(_stackprof.tag_strings[n]));
+            free(_stackprof.tag_strings[n]);
+        }
+        rb_hash_aset(results, sym_tag_strings, tag_strings);
+        free(_stackprof.tag_strings);
+        _stackprof.tag_strings = NULL;
+        _stackprof.tag_strings_len = 0;
+        _stackprof.tag_strings_capa = 0;
+    }
+
+    if (_stackprof.tag_string_table) {
+        st_free_table(_stackprof.tag_string_table);
+        _stackprof.tag_string_table = NULL;
+    }
+
+    // NOTE - it may be possible that there could be buffered samples and tags
+    // that were not captured but not accounted for in the report
+    if (_stackprof.sample_tags_len > 0) {
+        sample_tags = rb_ary_new_capa(_stackprof.sample_tags_len);
+        for (size_t n = 0; n < _stackprof.sample_tags_len; n++) {
+            VALUE tags = rb_hash_new();
+            st_foreach(_stackprof.sample_tags[n].tags, sample_tags_i, (st_data_t)tags);
+            rb_ary_push(sample_tags, tags);
+            rb_ary_push(sample_tags, ULONG2NUM(_stackprof.sample_tags[n].repeats));
+        }
+        rb_hash_aset(results, sym_sample_tags, sample_tags);
+    }
+
+    free(_stackprof.sample_tags);
+    _stackprof.record_tags = 0;
+    _stackprof.sample_tags = NULL;
+    _stackprof.sample_tags_len = 0;
+    _stackprof.sample_tags_capa = 0;
+    _stackprof.tag_source = Qnil;
+    _stackprof.tags = Qnil;
+    _stackprof.tag_thread_id = Qfalse;
+    _stackprof.tag_fiber_id = Qfalse;
+    _stackprof.last_tagset_matches = 0;
+    _stackprof.current_ruby_thread_id = 0;
+    _stackprof.overall_tags = 0;
+    _stackprof.buffered_tagsets = 0;
+    _stackprof.buffer_count = 0;
```

Comment:
perhaps I should remove this from the PR, I filed #207 with more details, but I encountered this rare case where old frames can stick around from the previous run.

I guess setting this to 0 here races with the postponed job that may also try to decrement it